src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs

```diff
@@ -0,0 +1,167 @@
+
+using System;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen
+{
+    public class CodeGenTupleTests : CSharpTestBase
+    {
+
+
+        private static readonly string trivial2uple =
+                    @"
+
+// PROTOTYPE: put in correct namespace
+namespace System.Runtime.CompilerServices
+{
+    // struct with two values
+    public struct ValueTuple<T1, T2>
+    {
+        public T1 Item1;
+        public T2 Item2;
+
+        public ValueTuple(T1 item1, T2 item2)
+        {
+            this.Item1 = item1;
+            this.Item2 = item2;
+        }
+
+        public override string ToString()
+        {
+            return '{' + Item1?.ToString() + "", "" + Item2.ToString() + '}';
+        }
+    }
+}
+
+            ";
+
+        [Fact]
+        public void SimpleTuple()
+        {
+            var source = @"
+class C
+{
+    static void Main()
+    {
+        var x = (1, 2);
+        System.Console.WriteLine(x.ToString());
+    }
+}
+" + trivial2uple;
+
+            var comp = CompileAndVerify(source, expectedOutput: "{1, 2}");
+            comp.VerifyDiagnostics();
+            comp.VerifyIL("C.Main", @"
+{
+  // Code size       28 (0x1c)
+  .maxstack  3
+  .locals init (System.Runtime.CompilerServices.ValueTuple<int, int> V_0) //x
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  ldc.i4.1
+  IL_0003:  ldc.i4.2
+  IL_0004:  call       ""System.Runtime.CompilerServices.ValueTuple<int, int>..ctor(int, int)""
+  IL_0009:  ldloca.s   V_0
+  IL_000b:  constrained. ""System.Runtime.CompilerServices.ValueTuple<int, int>""
+  IL_0011:  callvirt   ""string object.ToString()""
+  IL_0016:  call       ""void System.Console.WriteLine(string)""
+  IL_001b:  ret
+}");
+        }
+
+        [Fact]
+        public void SimpleTupleNested()
+        {
+            var source = @"
+class C
+{
+    static void Main()
+    {
+        var x = (1, (2, (3, 4)).ToString());
+        System.Console.WriteLine(x.ToString());
+    }
+}
+" + trivial2uple;
+
+            var comp = CompileAndVerify(source, expectedOutput: "{1, {2, {3, 4}}}");
+            comp.VerifyDiagnostics();
+            comp.VerifyIL("C.Main", @"
+{
+  // Code size       54 (0x36)
+  .maxstack  5
+  .locals init (System.Runtime.CompilerServices.ValueTuple<int, string> V_0, //x
+                System.Runtime.CompilerServices.ValueTuple<int, > V_1)
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  ldc.i4.1
+  IL_0003:  ldc.i4.2
+  IL_0004:  ldc.i4.3
+  IL_0005:  ldc.i4.4
+  IL_0006:  newobj     ""System.Runtime.CompilerServices.ValueTuple<int, int>..ctor(int, int)""
+  IL_000b:  newobj     ""System.Runtime.CompilerServices.ValueTuple<int, >..ctor(int, )""
```

Comment:
Something seems off with visualization of nested tuples (the type in actual IL seems correct). 
Will look at this in a later change.